### PR TITLE
rafthttp: check decode size before buffer alloc

### PR DESCRIPTION
**Pull request by gyuho**
Fix https://github.com/coreos/etcd/issues/5386.

### CODE REVIEW

File: rafthttp/msg_codec.go [hunk 1]
```diff
@@ -41,12 +42,20 @@ type messageDecoder struct {
 	r io.Reader
 }
 
+var (
+	readBytesLimit uint64 = 512 * 1024 // 512 MB
+	ErrReadLimit          = errors.New("rafthttp: error limit exceeded")
```
Comment (xiang90) [inline rafthttp/msg_codec.go]: ErrExceedSizeLimit

**Comment by xiang90** [comment]
lgtm

**Comment by gyuho** [comment]
Fixed. Merging in greens. Thanks.

File: rafthttp/msg_codec.go [hunk 2]
```diff
@@ -41,12 +42,20 @@ type messageDecoder struct {
 	r io.Reader
 }
 
+var (
+	readBytesLimit     uint64 = 512 * 1024 // 512 MB
```
Comment (nekto0n) [inline rafthttp/msg_codec.go]: That seems to me like 512 KB, am I counting wrong?

Comment (gyuho) [inline rafthttp/msg_codec.go]: oops, @nekto0n Do you want to send a patch to fix it? Thanks!

Comment (nekto0n) [inline rafthttp/msg_codec.go]: Sure, give me a sec.

Comment (xiang90) [inline rafthttp/msg_codec.go]: @nekto0n  We cannot do that for now. There is a 4MB limit on message size, 512KB is clearly not large enough. We set this to 512MB mainly for avoiding corrupted data to put down the server by asking it to allocate very large array.

**Pull request merged**